File: j-backend/src/models/StockMovement.js
/**
 * StockMovement model - Tracks all inventory changes (ledger).
 */
const { DataTypes } = require('sequelize');
const { sequelize } = require('../db');

const STOCK_MOVEMENT_REASONS = ['sale', 'purchase', 'adjustment', 'return', 'correction', 'transfer'];

const StockMovement = sequelize.define('StockMovement', {
  id: {
    type: DataTypes.UUID,
    defaultValue: DataTypes.UUIDV4,
    primaryKey: true,
  },
  shop_id: {
    type: DataTypes.UUID,
    allowNull: false,
    references: {
      model: 'shops',
      key: 'id',
    },
  },
  product_id: {
    type: DataTypes.UUID,
    allowNull: false,
    references: {
      model: 'products',
      key: 'id',
    },
  },
  change_qty: {
    type: DataTypes.INTEGER,
    allowNull: false,
  },
  reason: {
    type: DataTypes.ENUM(...STOCK_MOVEMENT_REASONS),
    allowNull: false,
  },
  reference_type: {
    type: DataTypes.TEXT,
    allowNull: true,
  },
  reference_id: {
    type: DataTypes.UUID,
    allowNull: true,
  },
  created_by: {
    type: DataTypes.UUID,
    allowNull: true,
    references: {
      model: 'users',
      key: 'id',
    },
  },
  meta_data: {
    type: DataTypes.JSONB,
    allowNull: false,
    defaultValue: {},
  },
}, {
  tableName: 'stock_movements',
  timestamps: true,
  createdAt: 'created_at',
  updatedAt: false,
});

StockMovement.STOCK_MOVEMENT_REASONS = STOCK_MOVEMENT_REASONS;

module.exports = StockMovement;

File: backend/app/models/stock_movement.py
# backend/app/models/stock_movement.py
import uuid
from sqlalchemy import Column, Integer, Text, text
from sqlalchemy.dialects.postgresql import UUID, JSONB
from sqlalchemy.sql import func
from sqlalchemy import TIMESTAMP
from sqlalchemy.orm import relationship
from sqlalchemy import ForeignKey
from sqlalchemy import Enum as SAEnum

from app.db.base import Base

stock_movement_reason = SAEnum('sale','purchase','adjustment','return','correction','transfer', name="stock_movement_reason")


class StockMovement(Base):
    __tablename__ = "stock_movements"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4, server_default=text("gen_random_uuid()"))
    shop_id = Column(UUID(as_uuid=True), ForeignKey("shops.id"), nullable=False)
    product_id = Column(UUID(as_uuid=True), ForeignKey("products.id"), nullable=False)
    change_qty = Column(Integer, nullable=False)
    reason = Column(stock_movement_reason, nullable=False)
    reference_type = Column(Text, nullable=True)
    reference_id = Column(UUID(as_uuid=True), nullable=True)
    created_by = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True)
    created_at = Column(TIMESTAMP(timezone=True), server_default=func.now(), nullable=False)
    meta_data = Column(JSONB, nullable=False, server_default=text("'{}'::jsonb"))  # Renamed from 'metadata' (SQLAlchemy reserved)

    # relationships
    product = relationship("Product", back_populates="stock_movements")
    shop = relationship("Shop", back_populates="stock_movements")
    created_by_user = relationship("User", back_populates="stock_movements")
